FEATURE/ElvaRivle/cIq5c09f

ElvaRivle opened the pull request:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * fixed PR review
--------------------------------------------------------------------
 OTHER
 * don't return entire collection
 * added new error message for page size not provided
--------------------------------------------------------------------
 TODO
 * /
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * added all fields to User model
 * added validation for all those fields
--------------------------------------------------------------------
 OTHER
 * added departments in `contants.js` file
 * seeding default users with all necessary attributes
--------------------------------------------------------------------
 TODO
 * add routes for GET all, GET one by ID, PATCH one by ID and DELETE one by ID
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * added route for getting all users
--------------------------------------------------------------------
 OTHER
 * /
--------------------------------------------------------------------
 TODO
 * add routes for GET one by ID, PATCH one by ID and DELETE one by ID
--------------------------------------------------------------------

becarselman commented on repositories/user.js:
```diff
@@ -5,9 +5,22 @@ async function getUserByEmail(email) {
+  } = { ...userData }
 
-  return await User.create({email: email, password: password, role: role})
+  return await User.create({email, name, surname, department, salary, stack, password, role})
```
you can simply pass userData here. You don't need to destruct everything and than pas separatly

becarselman commented on repositories/user.js:
```diff
@@ -5,9 +5,22 @@ async function getUserByEmail(email) {
+}
+
+async function getAllUsers() {
+  return await User.find({}).exec()
```
Prepare here for pagination. Because getAllUsers will return all n number of users.
You will need skip and limit.

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * added route for getting user by id
--------------------------------------------------------------------
 OTHER
 * created custom error class to specify custom fields in error (response code mostly)
 * added descriptive error messages for missing/wrong id
--------------------------------------------------------------------
 TODO
 * add routes for PATCH one by ID and DELETE one by ID
 * fix previous PR review
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * fixes #21
 * added pagination
 * moved pagination validation to middleware
--------------------------------------------------------------------
 OTHER
 * fixed issues from previous review
 * pagination attributes sent via query params, users' id sent via route params
--------------------------------------------------------------------
 TODO
 * add routes for PATCH one by ID and DELETE one by ID
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * added DELETE by id route
--------------------------------------------------------------------
 OTHER
 * /
--------------------------------------------------------------------
 TODO
 * add routes for PATCH one by ID
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * added route for PATCH one user by ID
--------------------------------------------------------------------
 OTHER
 * added validation rules for update, similar to register, but all fields are optional
--------------------------------------------------------------------
 TODO
 * /
--------------------------------------------------------------------

becarselman commented on repositories/user.js:
```diff
@@ -5,22 +5,41 @@ async function getUserByEmail(email) {
-  return await User.create({email: email, password: password, role: role})
+async function getAllUsers(page, pageSize) {
+  return await User
+      .find({})
```
I think you can remove {}

becarselman commented on middleware/users.js:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    //if page size is zero, mongoose will get entire collection
+    if (req.query.size < 0) {
```
I think we shouldn't get all collection in any case. What if we have 10k rows. Our backend will fall down.

ElvaRivle committed:
FEATURE/ElvaRivle/cIq5c09f

--------------------------------------------------------------------
 MAIN
 * fixed PR review
--------------------------------------------------------------------
 OTHER
 * don't return entire collection
 * added new error message for page size not provided
--------------------------------------------------------------------
 TODO
 * /
--------------------------------------------------------------------
